coalib/collecting/Importers.py

```diff
@@ -1,34 +1,30 @@
 from contextlib import ExitStack
+from importlib.util import spec_from_file_location
 import inspect
 import os
 import platform
-import sys
 
 from coala_utils.ContextManagers import suppress_stdout
 from coala_utils.decorators import arguments_to_lists, yield_once
 
+# in Python > 3.4 this can be imported from importlib.utils
+from coalib.misc.Compatibility import module_from_spec
+
 
 def _import_module(file_path):
     if not os.path.exists(file_path):
-        raise ImportError
+        raise ImportError("can's import non-existing file %s"
+                          % repr(file_path))
 
     module_name = os.path.splitext(os.path.basename(file_path))[0]
-    module_dir = os.path.dirname(file_path)
-
-    if module_dir not in sys.path:
-        sys.path.insert(0, module_dir)
-
-    # Ugly inconsistency: Python will insist on correctly cased module names
-    # independent of whether the OS is case-sensitive or not.
-    # We want all cases to match though.
-    if platform.system() == 'Windows':  # pragma: nocover
-        for cased_file_path in os.listdir(module_dir):
-            cased_module_name = os.path.splitext(cased_file_path)[0]
-            if cased_module_name.lower() == module_name.lower():
-                module_name = cased_module_name
-                break
-
-    return __import__(module_name)
+    spec = spec_from_file_location(module_name, file_path)
+    if not spec:
+        raise ImportError("%s doesn't seem to be a python module"
+                          % repr(file_path))
+
+    module = module_from_spec(spec)
+    spec.loader.exec_module(module)
+    return module
```

Comment:
@jayvdb Thanks for those hints! I see now that the `Importers._is_defined_in` check is the actual problem. We need to change this verification mechanism. Doing `sys.path.insert` and adding every imported module to `sys.modules` is a bad strategy. When all bears are discovered, we recursively load every `*.py` file in the search path, not only the actual bear modules. All those get added to `sys.modules`, even if we don't need them, even the `__init__.py` modules. This is clutter. In the end there is `sys.modules['__init__']`, referencing the last one loaded during the process:

```python
>>> sys.modules['__init__']
<module '__init__' from 'c:\\users\\zimmermann\\projects\\coala\\bears\\bears\\__init__.py'>
```

An having the modules' directories in `sys.path` also makes all non-bear `*.py` files directly `import`able even if not in `sys.modules`, also `import __init__` then always works